feat(variant): add new endpoint to clear snapshots

maugde opened the pull request:
Adding the new endpoint "clear-snapshots" will allow admin user to clear old snapshots in order to easily gain disk usage.

Commit:
feat(variant): add new endpoint to clear snapshots

Commit:
style(variant): reformat code

mabw-rte committed:
fix(variant): resolve the `Python-Lint` issues

mabw-rte commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1051,35 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
             )
             return False
+
+    def clear_all_snapshots(self, limit: int, params: t.Optional[RequestParameters] = None) -> None:
```
For `Mypy` you should deal first with the case `params=None` (it is an Optional parameter)

mabw-rte replied:
Add some Unit tests for this function.

mabw-rte commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1051,35 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+            HTTPException
+        """
+        if limit < 0:
+            raise HTTPException(status_code=400, detail=f"Limit cannot be negative (limit={limit})")
```
The Exception should be a `422` instead of `400`. You can look for a suitable Exception in `antarest/core/exceptions.py`.
If you think none of the `422` is ok for this case, you might create your own.

sylvlecl replied:
Note that it's also possible to use a `Query` annotation on the limit argument, with "ge=0" argument to specify that it should be greater or equal to 0.

mabw-rte commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1051,35 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+                ):
+                    self.clear_snapshot(variant)
+        else:
+            raise UserHasNotPermissionError
```
You should raise an instance of the class. I am not sure but that should be `UserHasNotPermissionError()`.

mabw-rte commented on antarest/study/web/variant_blueprint.py:
```diff
@@ -421,4 +422,32 @@ def create_from_variant(
         params = RequestParameters(user=current_user)
         raise NotImplementedError()
 
+    @bp.put(
```
You should add some integration tests for this endpoint

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1051,35 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+                )
+            )
+            for variant in result:
+                if variant.updated_at < datetime.now() - timedelta(hours=limit) or (
```
Maybe it's better to have an "and" condition: if it has been modified or read recently, we don't delete it.
For example, if it has been read recently, although it has not been modified, it makes sense to keep it.

sylvlecl commented on antarest/study/web/variant_blueprint.py:
```diff
@@ -421,4 +422,32 @@ def create_from_variant(
+            }
+        },
+    )
+    def clear_variant_snapshots(
```
After more thinking:
it could be relevant to execute this cleanup as a `Task` by the task service, because it will take a long time to execute.

Then here we should return the task ID. You can for example look at the `THERMAL_CLUSTER_SERIES_GENERATION` that Martin added recently, to see how it's done! it's quite simple you need to add a new task type and provide the function to be executed to the task service, instead of executing it directly in the variant service.

sylvlecl reviewed (commented):
See comment: the best option will be to execute the function as a task because it's a long processing

laurent-laporte-pro committed:
build(python): bump main project dependencies  (#1728)

bumps main projection dependencies:
- `pydantic` from 1.9 to 2.8: huge breaking change but with large
   performance benefits expected on serialization
- `fastapi` from 0.73 to 0.110
- `uvicorn` from to 0.15 to 0.30
- `mypy` from 1.4 to 1.11

It also brings few changes inside dependencies
- Drop `requests` in favor of `httpx`
- Drop `fastapi-jwt-auth` as they do not and will not support pydantic
v2. We've decided to fork their code and adapt it as it's really light
(see new folder `/antarest/fastapi_jwt_auth`)

These changes also induced other minor dependencies bump: `jinja2`,
`typing_extensions`, `PyJWT`, `python-multipart`

Last, this work includes fixes on the API prefix addition when
running in standalone mode (desktop version). We now distinguish
properties root_path (used when behind a proxy) and api_prefix
(which actually makes our server add a prefix).

Co-authored-by: belthlemar <martin.belthle@rte-france.com>
Co-authored-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

laurent-laporte-pro committed:
build(python): bump main project dependencies  (#1728)

bumps main projection dependencies:
- `pydantic` from 1.9 to 2.8: huge breaking change but with large
   performance benefits expected on serialization
- `fastapi` from 0.73 to 0.110
- `uvicorn` from to 0.15 to 0.30
- `mypy` from 1.4 to 1.11

It also brings few changes inside dependencies
- Drop `requests` in favor of `httpx`
- Drop `fastapi-jwt-auth` as they do not and will not support pydantic
v2. We've decided to fork their code and adapt it as it's really light
(see new folder `/antarest/fastapi_jwt_auth`)

These changes also induced other minor dependencies bump: `jinja2`,
`typing_extensions`, `PyJWT`, `python-multipart`

Last, this work includes fixes on the API prefix addition when
running in standalone mode (desktop version). We now distinguish
properties root_path (used when behind a proxy) and api_prefix
(which actually makes our server add a prefix).

Co-authored-by: belthlemar <martin.belthle@rte-france.com>
Co-authored-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

hdinia committed:
feat(ui-common): integrate `GlideDataGrid` into `MatrixGrid` (#2134)

Commit:
ci(tests): add unit and integration tests

Commit:
fix(tests): add context and clean test

MartinBelthle committed:
feat(pydantic): use pydantic serialization (#2139)


This will allow to benefit from improved serialization performances
of pydantic v2, even for non-pydantic classes.

fix [ANT-2052]

sylvlecl committed:
fix(ci): multiply timeouts on windows platform (#2137)


The aim is to solve irrelevant failures on windows CI.
It often fails when waiting for study upgrades, in particular.
The PR re-activates the use of more workers (logical cores
instead of CPUs).

ANT-2043

---------

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

MartinBelthle committed:
chore(front-end): add license headers inside front-end (#2145)


Fix [ANT-1916] for the front-end part

MartinBelthle committed:
chore(variants): increase timeout duration for variant generation (#2144)

Commit:
feat(variantstudy): reformat snapshot clear task into task

MartinBelthle committed:
style(license): reformat license header inside front-end (#2148)

skamril committed:
chore(license): add a new ESLint rule to check license header (#2150)

Commit:
fix(variantstudy): add typing and reformat code

Commit:
feat(tests.integration): add user rights test

Commit:
fix(variantstudy): sort imports

Commit:
style(tasks.models): reformat code

Commit:
style(test_variant_manager): fix window test

sylvlecl committed:
ci(tests): reduce number of workers for tests (#2149)

The use of as many workers as logical cores
seems to cause significant increases of some
monitored durations, including on Linux, which
causes irrelevant CI failures.

We revert to using as many workers as actual CPUs.

Also, load balancing of tests is improved using worksteal,
which works far better with our tests which have very different
execution times (integration tests are far longer than unit tests).

---------

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

mabw-rte committed:
feat(aggregation-api): delete index from the response file  (#2151)

MartinBelthle committed:
fix(playlist): change response model to accept optional answers (#2152)

MartinBelthle committed:
style(api): change apidoc example to make it work (#2155)


[ANT-1993]

sylvlecl commented on tests/integration/variant_blueprint/test_variant_manager.py:
```diff
@@ -45,6 +47,60 @@ def variant_id_fixture(
+
+                # Generate snapshot for each variant
+                client.put(f"/v1/studies/{variant_ids[index]}/generate", headers=admin_headers)
+        time.sleep(0.1)  # wait for the filesystem to be updated
```
You can use the `wait_task_completion` method to make sure that the generation task has ended (you will need to get the id of the task from the `PUT` call above).

sylvlecl commented on tests/study/storage/variantstudy/test_variant_study_service.py:
```diff
@@ -239,3 +242,168 @@ def test_generate_task(
+
+        # Begin tests
+        # A user without rights cannot clear snapshots
+        try:
```
you may use `with pytest.raises(UserHasNotPermissionError):` instead of try/catch, to check that this exception is raised.

sylvlecl commented on tests/integration/variant_blueprint/test_variant_manager.py:
```diff
@@ -330,3 +386,29 @@ def test_outputs(client: TestClient, admin_access_token: str, variant_id: str, t
+    assert res.status_code == 422
+    assert res.json().get("exception") == "VariantAgeMustBePositive"
+
+    assert older.exists() and old.exists() and recent.exists()
```
would be nice to check also a valid call, not only the invalid value.

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1054,90 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
             )
             return False
+
+    def clear_all_snapshots(self, limit: int, params: t.Optional[RequestParameters]) -> str:
```
for clarity, we could rename `limit` to `retention_time` and change the type to `datetime.timedelta` in order to be independent from the unit used (hours).
The conversion int --> timedelta will find a good home in the web layer, which translates request parameters to service arguments.

sylvlecl commented on antarest/study/web/variant_blueprint.py:
```diff
@@ -421,4 +421,32 @@ def create_from_variant(
+        },
+    )
+    def clear_variant_snapshots(
+        limit: int = 24,
```
For clarity in the API we can name it `retention_hours`

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1054,90 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+                if variant.updated_at < datetime.now() - timedelta(hours=self._limit):
+                    if variant.last_access and variant.last_access < datetime.now() - timedelta(hours=self._limit):
+                        logger.info(f"Variant {variant.id} detected.")
+                        path = self._variant_study_service.get_study_path(variant)
```
we'd better use the existing method `clear_snapshot`, which in particular also invalidates the cache of the study

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1054,90 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+        study_tasks = self.task_service.list_tasks(
+            TaskListFilter(
+                ref_id="snapshot_cleaning",
+                type=[TaskType.THERMAL_CLUSTER_SERIES_GENERATION],
```
wrong type here

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1054,90 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+            logger.info(msg)
+            return msg
+
+        # Make sure the task is not running
```
Actually, I don't think we need to check that another clean is running.
It's ok to run multiple cleaning independently, there is no real reason to forbid it.

At worse, some variants will have their snapshot cleared multiple times, which will just do nothing after the first time.

sylvlecl commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1050,3 +1054,90 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+            UserHasNotPermissionError
+            VariantAgeMustBePositive
+        """
+        if params is None:
```
we should also check that `user.is_site_admin()`, to allow only admins to use this feature

sylvlecl reviewed (changes requested):
A few changes required:
- check that user is admin
- not necessary to check for already running tasks
- advise on naming in particular for the API

Also the installer submodule is updated: it should not be the case in this PR.

MartinBelthle committed:
fix(api): allow `nominalcapacity` to be a float inside API response (#2158)


It also removes useless xpansion code that was unreachable

Fix [ANT-2214]

MartinBelthle committed:
fix(adq_patch): set default value for field `enable-first-step` to False (#2160)


Fix [ANT-2218]

Commit:
fix(test_variant_manager): run snapshot clearing as a task, update tests, improve code style

Commit:
style(test_variant_study_service): reorder imports

Commit:
style(variant_study_service): add default value for `params` parameter

Commit:
style(variant_study_service): update type test

Commit:
style(variant_study_service): remove duplicate import

Commit:
Merge branch 'refs/heads/dev' into feature/ANT-1987-add-snapshot-endpoint

# Conflicts:
#	antarest/study/storage/variantstudy/variant_study_service.py
#	sonar-project.properties
#	webapp/package-lock.json
#	webapp/src/components/App/Singlestudy/explore/Debug/Data/Json.tsx
#	webapp/src/components/App/Singlestudy/explore/Debug/Data/Matrix.tsx
#	webapp/src/components/App/Singlestudy/explore/Debug/Data/Text.tsx
#	webapp/src/components/App/Singlestudy/explore/Debug/Data/style.ts
#	webapp/src/components/App/Singlestudy/explore/Debug/DebugContext.ts
#	webapp/src/components/App/Singlestudy/explore/Debug/Tree/index.tsx
#	webapp/src/components/App/Studies/StudyTree.tsx
#	webapp/src/components/common/DownloadMatrixButton.tsx
#	webapp/src/components/common/Fieldset.tsx
#	webapp/src/components/common/JSONEditor/index.tsx
#	webapp/src/components/common/MatrixGrid/MatrixActions.tsx
#	webapp/src/components/common/MatrixGrid/useMatrix.test.tsx
#	webapp/src/components/common/MatrixGrid/useMatrix.ts
#	webapp/src/services/api/studies/raw/types.ts

MartinBelthle commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1053,3 +1055,68 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+        Raises:
+            UserHasNotPermissionError
+        """
+        if params is None or (params.user and not params.user.is_site_admin()):
```
You should add `or user.is_admin_token()` to support the fact that an admin calls this endpoint via a script

MartinBelthle commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1053,3 +1055,68 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+                )
+            )
+            for variant in variant_list:
+                if variant.updated_at < datetime.now() - self._retention_hours:
```
This comment goes for this line and the following:

Not sure and you've probably tested it inside the tests but there could be a problem of timezone UTC, we should check how we save the field info to be sure it's not a problem

Other remark, are we sure we'll always have the field ? Cause line 1109 you don't check if you have it and line 1110 you check it.

maugde replied:
I indeed checked only on the `last_access` field because it seemed that the `updated_at` field was always set after a variant creation. In order to remain consistant, i'll add the same test as `last_access`.

MartinBelthle commented on antarest/study/storage/variantstudy/variant_study_service.py:
```diff
@@ -1053,3 +1055,68 @@ def initialize_additional_data(self, variant_study: VariantStudy) -> bool:
+            for variant in variant_list:
+                if variant.updated_at < datetime.now() - self._retention_hours:
+                    if variant.last_access and variant.last_access < datetime.now() - self._retention_hours:
+                        logger.info(f"Variant {variant.id} detected.")
```
Not a huge fan of this log as it's not really precise. IMO we should remove it as the log inside the clear_snapshot method is enough

MartinBelthle commented on antarest/study/web/variant_blueprint.py:
```diff
@@ -416,4 +416,33 @@ def create_from_variant(
+
+        Args: limit (int, optional): Number of hours to clear. Defaults to 24.
+
+        Returns: None
```
It doesn't return None it returns the created task id as you typed it above.

MartinBelthle commented on antarest/study/web/variant_blueprint.py:
```diff
@@ -416,4 +416,33 @@ def create_from_variant(
+        """
+        retention_hours = datetime.timedelta(hours=hours)
+        logger.info(
+            f"Delete all variant snapshots older than {retention_hours} hours (24 by default)",
```
Don't really see why you add the (24 by default) inside the log as it doesn't bring any info IMO

MartinBelthle commented on tests/integration/variant_blueprint/test_variant_manager.py:
```diff
@@ -45,6 +48,66 @@ def variant_id_fixture(
+) -> [str]:
+    """Generate some snapshots with different date of update and last access"""
+
+    class FakeDatetime:
```
nice way to do this

MartinBelthle commented on tests/study/storage/variantstudy/test_variant_study_service.py:
```diff
@@ -239,3 +242,181 @@ def test_generate_task(
+            assert variant.is_dir()
+            assert list(variant.iterdir())[0].name == "snapshot"
+
+        # Begin tests
```
Perhaps just separate the set up and the tests with a bigger sign to ensure better readability. In other tests, we used something like this:
```python
# =============================
#  SET UP
# ============================= 
```

MartinBelthle commented on tests/study/storage/variantstudy/test_variant_study_service.py:
```diff
@@ -239,3 +242,181 @@ def test_generate_task(
+        )
+        variant_study_service.task_service.await_task(task_id)
+
+        # Check if all snapshots was cleared
```
were

MartinBelthle commented on tests/study/storage/variantstudy/test_variant_study_service.py:
```diff
@@ -239,3 +242,181 @@ def test_generate_task(
+        # Check if all snapshots was cleared
+        nb_snapshot_dir = 0  # after the for iterations, must equal 0
+        for variant_path in variant_study_path.iterdir():
+            if variant_path.joinpath("snapshot").exists():
```
Better way to do it is `assert not variant_path.joinpath("snapshot").exists()` instead of introducing a counter and you can even do it inside one line

MartinBelthle commented on tests/study/storage/variantstudy/test_variant_study_service.py:
```diff
@@ -239,3 +242,181 @@ def test_generate_task(
+        )
+        variant_study_service.task_service.await_task(task_id)
+
+        # Check if old snapshots was successfully cleared
```
were

MartinBelthle reviewed (commented):
Seems good, just minor changes/questions , you've added many tests it's nice :)

Commit:
chore(snapshot_cleaning): fix minor datetime issues, add annotations, update comments

Commit:
chore(variant_study_service): add condition in variant selection
